File: src/examples/eina/eina_array_01.c
//Compile with:
//gcc -g eina_array_01.c -o eina_array_01 `pkg-config --cflags --libs eina`

#include <stdio.h>
#include <string.h>

#include <Eina.h>

static Eina_Bool
_print(const void *container EINA_UNUSED, void *data, void *fdata EINA_UNUSED)
{
  printf("%s\n", (char *)data);
   return EINA_TRUE;
}

int
main(int argc EINA_UNUSED, char **argv EINA_UNUSED)
{
   const char* strings[] = {
      "helo", "hera", "starbuck", "kat", "boomer",
      "hotdog", "longshot", "jammer", "crashdown", "hardball",
      "duck", "racetrack", "apolo", "husker", "freaker",
      "skulls", "bulldog", "flat top", "hammerhead", "gonzo"
   };
   Eina_Array *array;
   unsigned int i;

   eina_init();

   array = eina_array_new(10);
   eina_array_step_set(array, sizeof(*array), 20);

   for (i = 0; i < 20; i++)
     eina_array_push(array, strdup(strings[i]));

   printf("array count: %d\n", eina_array_count(array));
   eina_array_foreach(array, _print, NULL);

   printf("Top gun: %s\n", (char*)eina_array_data_get(array, 2));

   while (eina_array_count(array))
     free(eina_array_pop(array));

   eina_array_free(array);

   eina_shutdown();

   return 0;
}

File: src/examples/eina/EinaArray01.cs
/*
 * Copyright 2019 by its authors. See AUTHORS.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using static System.Console;

public class ExampleEinaArray01
{
    public static void Main()
    {
        Eina.Config.Init();

        var strings = new string[]{
            "helo", "hera", "starbuck", "kat", "boomer",
            "hotdog", "longshot", "jammer", "crashdown", "hardball",
            "duck", "racetrack", "apolo", "husker", "freaker",
            "skulls", "bulldog", "flat top", "hammerhead", "gonzo"
        };

        var array = new Eina.Array<string>(20U);

        // Push new elements
        foreach (string s in strings)
        {
            WriteLine("push: " + s);
            array.Push(s);
        }

        // Check count
        WriteLine("array count: " + array.Count());

        // Iterate over the array
        int idx = 0;
        foreach (string s in array)
        {
            WriteLine($"at[{idx}]: {s}");
            ++idx;
        }

        // Remove one by one
        while (array.Length != 0)
            WriteLine("pop: " + array.Pop());
    }
}

